src/function/gds/K_Core_Decomposition.cpp

```diff
@@ -0,0 +1,294 @@
+        : FrontierPair(curFrontier, nextFrontier, maxThreads), numNodesMap{numNodesMap} {
+        for (const auto& [tableID, curNumNodes] : numNodesMap) {
+            curVertexValues.allocate(tableID, curNumNodes, mm);
+            auto data = curVertexValues.getData(tableID);
+            for (auto i = 0u; i < curNumNodes; ++i) {
+                data[i].store(0, std::memory_order_relaxed);
+            }
+        }
+    }
+
+    void initRJFromSource(common::nodeID_t /* source */) override{};
+
+    void pinCurrFrontier(common::table_id_t tableID) override {
+        FrontierPair::pinCurrFrontier(tableID);
+        curDenseFrontier->ptrCast<PathLengths>()->pinCurFrontierTableID(tableID);
+    }
+
+    void pinNextFrontier(common::table_id_t tableID) override {
+        FrontierPair::pinNextFrontier(tableID);
+        nextDenseFrontier->ptrCast<PathLengths>()->pinNextFrontierTableID(tableID);
+    }
+
+    void pinVertexValues(common::table_id_t tableID) {
+        vertexValues = curVertexValues.getData(tableID);
+    }
+
+    void beginFrontierComputeBetweenTables(table_id_t curTableID, table_id_t nextTableID) override {
+        FrontierPair::beginFrontierComputeBetweenTables(curTableID, nextTableID);
+        pinVertexValues(nextTableID);
+    }
+
+    void beginNewIterationInternalNoLock() override {
+        std::swap(curDenseFrontier, nextDenseFrontier);
+        updateSmallestDegree();
+        curDenseFrontier->ptrCast<PathLengths>()->incrementCurIter();
+        nextDenseFrontier->ptrCast<PathLengths>()->incrementCurIter();
+    }
+
+    uint64_t addToVertexDegree(common::offset_t offset, uint64_t degreeToAdd) {
+        return vertexValues[offset].fetch_add(degreeToAdd, std::memory_order_relaxed);
+    }
+
+    // Called to remove degrees from a neighbouring vertex
+    // Returns whether the neighbouring vertex should be set as active or not
+    bool removeFromVertex(common::nodeID_t nodeID) {
+        int curSmallest = curSmallestDegree.load(std::memory_order_relaxed);
+        int curVertexDegree =
+            curVertexValues.getData(nodeID.tableID)[nodeID.offset].load(std::memory_order_relaxed);
+        // The vertex should be set as active if it will be considered in a future iteration
+        // The vertex should be set as inactive if it will be processed this iteration
+        if (curVertexDegree > curSmallest) {
+            curVertexValues.getData(nodeID.tableID)[nodeID.offset].fetch_sub(1,
+                std::memory_order_relaxed);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    void updateSmallestDegree() { curSmallestDegree.fetch_add(1, std::memory_order_relaxed); }
+
+    uint64_t getSmallestDegree() { return curSmallestDegree.load(std::memory_order_relaxed); }
+
+    uint64_t getVertexValue(common::offset_t offset) {
+        return vertexValues[offset].load(std::memory_order_relaxed);
+    }
+
+private:
+    bool updated = false;
```

Comment:
I have made a comment about this. Let's make sure all comments are resolved before you hold them